### ISSUE TITLE
Add StructuralComparer<T>

### ISSUE THREAD

**Issue by kzu** [dotnet/corefx#24205]
Quite a few APIs can take an `IEqualityComparer<T>` (most notably `HashSet<T>`, `Dictionary<T>` and `ConcurrentDictionary<T>` ) and having a structural comparer for those would be quite useful. In fact, the non-generic `StructuralComparisons.StructuralEqualityComparer` even already exists!

So how about adding the obvious dumb typed implementation on top of it? Pretty much:

```
public class StructuralComparer<T> : IEqualityComparer<T>
{
    public static IEqualityComparer<T> Default { get; } = new StructuralComparer<T>();

    public bool Equals(T x, T y) => StructuralComparisons.StructuralEqualityComparer.Equals(x, y);

    public int GetHashCode(T obj) => StructuralComparisons.StructuralEqualityComparer.GetHashCode(obj);
}
```

And that's all we need :)

**Comment by jkotas** [comment]
> So how about adding the obvious dumb typed implementation on top of it?

This would be pretty slow... .

**Comment by kzu** [comment]
Why would it? Note that the static property initializer will run only once. In other words, this succeeds:

```c#
	var first = StructuralComparer<string>.Default;
	var second = StructuralComparer<string>.Default;

	Debug.Assert(Object.ReferenceEquals(first, second));
```

So, unless the underlying `StructuralComparisons.StructuralEqualityComparer` is slow, there's no reason why this would be?

**Comment by jkotas** [comment]
Right, `StructuralComparisons.StructuralEqualityComparer` is not fast to start with it.

And StructuralEqualityComparer is non-generic and so `public bool Equals(T x, T y) => StructuralComparisons.StructuralEqualityComparer.Equals(x, y);` would box that would make it even slower.

**Comment by ianhays** [comment]
Definitely agree that the implementation above would be too slow. Could we instead just write a new `StructuralComparer<T>` that didn't call to the nongeneric `StructuralComparisons.StructuralEqualityComparer`?

That I think would be useful.
